src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,174 @@
+
+class ChunkedStreamDecoder implements ReadableStreamInterface
+{
+    const CRLF = "\r\n";
+    const MAX_BUFFER_LENGTH = 32768; // 32KB
+
+    use EventEmitterTrait;
+
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var int
+     */
+    protected $remainingLength = 0;
+
+    /**
+     * @var int
+     */
+    protected $maxBufferLength;
+
+    /**
+     * @var bool
+     */
+    protected $nextChunkIsLength = true;
+
+    /**
+     * @var ReadableStreamInterface
+     */
+    protected $stream;
+
+    /**
+     * @param ReadableStreamInterface $stream
+     * @param int $maxBufferLength
+     */
+    public function __construct(ReadableStreamInterface $stream, $maxBufferLength = self::MAX_BUFFER_LENGTH)
+    {
+        $this->stream = $stream;
+        $this->stream->on('data', array($this, 'handleData'));
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+        ]);
+        $this->maxBufferLength = $maxBufferLength;
+    }
+
+    /** @internal */
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $bufferLength = strlen($this->buffer);
+            $continue = $this->iterateBuffer();
+            $iteratedBufferLength = strlen($this->buffer);
+        } while (
+            $continue &&
+            $bufferLength !== $iteratedBufferLength &&
+            $iteratedBufferLength > 0 &&
+            strpos($this->buffer, static::CRLF) !== false &&
+            $iteratedBufferLength <= $this->maxBufferLength
+        );
+
+        if ($iteratedBufferLength >= $this->maxBufferLength) {
+            $this->emit('error', [
+                new Exception('The current buffer is longer then the ' . $this->maxBufferLength / 1024 . 'KB we have set as maximum'),
+            ]);
+            $this->close();
+            return false;
+        }
+    }
+
+    protected function iterateBuffer()
+    {
+        if (strlen($this->buffer) <= 1) {
+            return false;
+        }
+
+        if ($this->nextChunkIsLength) {
+            $crlfPosition = strpos($this->buffer, static::CRLF);
+            if ($crlfPosition === false) {
+                return false; // Chunk header hasn't completely come in yet
+            }
+            $this->nextChunkIsLength = false;
+            $lengthChunk = substr($this->buffer, 0, $crlfPosition);
+            if (strpos($lengthChunk, ';') !== false) {
+                list($lengthChunk) = explode(';', $lengthChunk, 2);
+            }
+            if (dechex(hexdec($lengthChunk)) !== $lengthChunk) {
+                $this->emit('error', [
+                    new Exception('Unable to validate "' . $lengthChunk . '" as chunk length header"'),
+                ]);
+                $this->close();
+                return false;
+            }
+            $this->remainingLength = hexdec($lengthChunk);
+            if ($this->remainingLength >= $this->maxBufferLength) {
```

Comment:
Perhaps I'm missing something really obvious here, but why would we want to limit the chunk length at all? After all, it should be up to the client to decide which kind of streams they accept? E.g. we also don't put a limit on the `Content-Length` header either and this doesn't really matter either because we end up with a stream instance anyway and won't keep any of this in any buffers.

Note that I've previously suggested (and thereafter retracted the suggestion) buffering the whole chunk before emitting, in this case we would certainly have needed to limit the maximum chunk length.